stl/inc/xutility

```diff
@@ -5767,6 +5768,121 @@ private:
     _Udiff _Bmask; // 2^_Bits - 1
 };
 
+template <class _Diff, class _Urng>
+class _Rng_from_urng_v2 { // wrap a URNG as an RNG
+public:
+    using _Ty0 = make_unsigned_t<_Diff>;
+    using _Ty1 = typename _Urng::result_type;
+
+    using _Udiff                              = conditional_t<sizeof(_Ty1) < sizeof(_Ty0), _Ty0, _Ty1>;
+    constexpr static unsigned int _Udiff_bits = sizeof(_Udiff) * CHAR_BIT;
+    using _Uprod = conditional_t<_Udiff_bits <= 16, uint32_t, conditional_t<_Udiff_bits <= 32, uint64_t, _Unsigned128>>;
+
+    explicit _Rng_from_urng_v2(_Urng& _Func) : _Ref(_Func) {}
+
+    _Diff operator()(_Diff _Index) { // adapt _Urng closed range to [0, _Index)
+        // From Daniel Lemire, "Fast Random Integer Generation in an Interval", ACM Trans. Model. Comput. Simul. 29 (1),
+        // 2019.
+        _Udiff _Mask;
+        unsigned int _Niter;
+
+        if constexpr (_Bits < _Udiff_bits) {
+            _Mask  = 0;
+            _Niter = 0;
+            do {
+                _Mask <<= _Bits;
+                _Mask |= _Bmask;
+                ++_Niter;
+            } while (_Mask < _Udiff(_Index - 1));
+        } else {
+            _Mask  = _Bmask;
+            _Niter = 1;
+        }
+
+        auto _Product = _Get_random_product(_Index, _Niter);
+        auto _Rem     = static_cast<_Udiff>(_Product) & _Mask;
+
+        if (_Rem < _Index) {
+            const auto _Threshold = (_Mask - (_Index - 1)) % _Index;
+            while (_Rem < _Threshold) {
+                _Product = _Get_random_product(_Index, _Niter);
+                _Rem     = static_cast<_Udiff>(_Product) & _Mask;
+            }
+        }
+
+        unsigned int _Generated_bits;
+        if constexpr (_Bits < _Udiff_bits) {
+            _Generated_bits = static_cast<unsigned int>(_Popcount(_Mask));
+        } else {
+            _Generated_bits = _Udiff_bits;
+        }
+
+        return static_cast<_Diff>(_Product >> _Generated_bits);
+    }
+
+    _Udiff _Get_all_bits() {
+        _Udiff _Ret = _Get_bits();
+
+        if constexpr (_Bits < _Udiff_bits) {
+            for (unsigned int _Num = _Bits; _Num < _Udiff_bits; _Num += _Bits) { // don't mask away any bits
+                _Ret <<= _Bits;
+                _Ret |= _Get_bits();
+            }
+        }
+
+        return _Ret;
+    }
+
+    _Rng_from_urng_v2(const _Rng_from_urng_v2&)            = delete;
+    _Rng_from_urng_v2& operator=(const _Rng_from_urng_v2&) = delete;
+
+private:
+    _Udiff _Get_bits() { // return a random value within [0, _Bmask]
+        for (;;) { // repeat until random value is in range
+            _Udiff _Val = _Ref() - (_Urng::min)();

```

Comment:
We can't do it unconditionally, because the TR1 URBG types aren't `constexpr`. We'd have to special-case it, like I did for `generate_canonical` in #2498.

Reply:
@StephanTLavavej I made this a `static constexpr` variable, then noticed your comment [here](https://github.com/microsoft/STL/pull/3024#discussion_r950622575) that `static` is unnecessary. FYI, I wrote it that way because they result in different debug codegen, for some reason: https://godbolt.org/z/he1f3hr9d.

Reply:
Interesting, thanks for checking! No objection to `static constexpr` then.